README.md
# rs-storage cookbook

[![Release](https://img.shields.io/github/release/rightscale-cookbooks/rs-storage.svg?style=flat)][release]
[![Build Status](https://img.shields.io/travis/rightscale-cookbooks/rs-storage.svg?style=flat)][travis]

[release]: https://github.com/rightscale-cookbooks/rs-storage/releases/latest
[travis]: https://travis-ci.org/rightscale-cookbooks/rs-storage

Provides recipes for managing volumes on a Server in a RightScale supported cloud which include:

* Creation of single and multi-stripe volumes
* Taking backups of volumes
* Restoring from backups
* Schedule periodic backups
* Detaching and deleting volumes when server is decommissioned.

Github Repository: [https://github.com/rightscale-cookbooks/rs-storage](https://github.com/rightscale-cookbooks/rs-storage)

# Requirements

* Requires Chef 12
* Requires Ruby 2.3.1
* Requires [RightLink 10](http://docs.rightscale.com/rl10/) See cookbook version 1.0.7 for RightLink 6 support
* Platform
  * Ubuntu 12.04, 14.04
  * CentOS 6, 7
* Cookbooks
  * [chef_handler](http://community.opscode.com/cookbooks/chef_handler)
  * [filesystem](http://community.opscode.com/cookbooks/filesystem)
  * [lvm](http://community.opscode.com/cookbooks/lvm)
  * [marker](http://community.opscode.com/cookbooks/marker)
  * [rightscale_volume](http://community.opscode.com/cookbooks/rightscale_volume)
  * [rightscale_backup](http://community.opscode.com/cookbooks/rightscale_backup)

# Usage

## Creating a new volume

To create a new volume, run the `rs-storage::volume` recipe with the following attributes set:

- `node['rs-storage']['device']['nickname']` - the nickname of the volume
- `node['rs-storage']['device']['volume_size']` - the size of the volume to create
- `node['rs-storage']['device']['filesystem']` - the filesystem to use on the volume
- `node['rs-storage']['device']['mount_point']` - the location to mount the volume

This will create a new volume, attach it to the server, format it with the filesystem specified, and mount it on the
location specified.

### Provisioned IOPS on EC2

To create a volume with IOPS on EC2, set the following attribute before running `rs-storage::volume` recipe:

- `node['rs-storage']['device']['iops']` - the value of IOPS to use

## Creating a logical volume with striping

To create a logical volume with striping using LVM, run the `rs-storage::stripe` recipe with the following attributes
set:

- `node['rs-storage']['device']['nickname']` - the nickname to use as prefix for the logical volume
- `node['rs-storage']['device']['count']` - number of volumes to create in the logical volume
- `node['rs-storage']['device']['volume_size']` - the total size of the logical volume
- `node['rs-storage']['device']['filesystem']` - the filesystem to use on the logical volume
- `node['rs-storage']['device']['mount_point']` - the location to mount the logical volume

This will create the number of volumes specified in `node['rs-storage']['device']['count']`. Each volume created
will have a nickname of `"#{nickname}_#{stripe_number}"`. The size for each volume is calculated by the following
formula:

```ruby
(total_size.to_f / device_count.to_f).ceil

# For example, total size = 10, stripe count = 3
(10.0 / 3.0).ceil
# => 4.0
```

This will create a volume group with the name `"#{nickname}-vg"` and a logical volume in it with the name
`"#{nickname}-lv"`, format it with the filesystem specified, and mount it on the location specified.

## Backing up volume(s) & Cleaning up backups

To create a backup of all volumes attached to the server, run the `rs-storage::backup` recipe with the following
attributes set:

- `node['rs-storage']['backup']['lineage']` - the lineage to be used for the backup

The backup process will create a snapshot of all volumes attached to the server (except the boot disk if there is one).
The backup recipe also handles the cleanup of old volume snapshots and accepts the following attributes:

- `node['rs-storage']['backup']['keep']['keep_last']` - number of last backups to keep from deleting
- `node['rs-storage']['backup']['keep']['dailies']` - number of daily backups to keep
- `node['rs-storage']['backup']['keep']['weeklies']` - number of weekly backups to keep
- `node['rs-storage']['backup']['keep']['monthlies']` - number of monthly backups to keep
- `node['rs-storage']['backup']['keep']['yearlies']` - number of yearly backups to keep

This will cleanup the old snapshots on the cloud based on the criteria specified.

## Restoring a volume from a backup

To restore a volume from backup, run the `rs-storage::volume` recipe with the same set of attributes mentioned in the
[previous section](#creating-a-new-volume) along with the following attribute:

- `node['rs-storage']['restore']['lineage']` - the lineage to restore the backup from

This will restore the volume from the backup instead of creating a new one. By default, the backup with the latest
timestamp will be restored. To restore a backup from a specific timestamp, set the following attribute:

- `node['rs-storage']['restore']['timestamp']` - the timestamp of the backup to restore from (in seconds since UNIX
  epoch)

## Restoring a logical volume composed of multiple volumes from a backup

To restore a logical volume composed of multiple volumes from a backup, run the `rs-storage::stripe` recipe with the
same set of attributes mentioned in the [previous section](#creating-stripe-of-volumes) along with the following
attribute:

- `node['rs-storage']['restore']['lineage']` - the lineage to restore the backup from

This will restore multiple volumes from the backup matching the lineage. By default, the backup with the latest
timestamp will be restored. To restore backup from a specific timestamp, set the following attribute:

- `node['rs-storage']['restore']['timestamp']` - the timestamp of the backup to restore from (in seconds since UNIX
  epoch)

## Scheduling automated backups of volume(s)

To schedule automated backups, run the `rs-storage::schedule` recipe with the following attributes set:

- `node['rs-storage']['schedule']['enable']` - to enable/disable automated backups
- `node['rs-storage']['schedule']['hour']` - the hour to take the backup on (should use crontab syntax)
- `node['rs-storage']['schedule']['minute']` - the minute to take the backup on (should use crontab syntax)
- `node['rs-storage']['backup']['lineage']` - the lineage name to be used for the backup

This will create a crontab entry to run the `rs-storage::backup` recipe periodically at the given minute and hour. To
disable the automated backups, simply set `node['rs-storage']['schedule']['enable']` to `false` and rerun the
`rs-storage::schedule` recipe and this will remove the crontab entry.

## Deleting volume(s)

This operation should be part of the decommission bundle in a RightScale ServerTemplate where the volumes attached to
the server are detached and deleted from the cloud but this can also be used as an operational recipe. This recipe will
do nothing in the following conditions:

- when the server enters a stop state
- when the server reboots

This recipe also has a safety attribute `node['rs-storage']['device']['destroy_on_decommission']`. This attribute will
be set to `false` by default and should be overridden and set to `true` in order for the devices to be detached and
deleted. If an LVM is found (created using `rs-storage::stripe`), the LVM will be conditionally removed before detaching
the volume.

# Attributes

- `node['rs-storage']['device']['nickname']` - The nickname for the device or the logical volume comprising multiple of
  devices. Default is `'data_storage'`.
- `node['rs-storage']['device']['mount_point']` - The mount point for the device. Default is `'/mnt/storage'`.
- `node['rs-storage']['device']['volume_size']` - The size (in gigabytes) of the volume to be created. If multiple
  devices are used, this will be the total size of the logical volume. Default is `10`.
- `node['rs-storage']['device']['count']` - The number of devices to be created for the logical volume. Default is `2`.
- `node['rs-storage']['device']['iops']` - The IOPS value to be used for EC2 Provisioned IOPS. This attribute should
  only be used with Amazon EC2. Default is `nil`.
- `node['rs-storage']['device']['filesystem']` - The filesystem to be used on the device. Default is `'ext4'`.
- `node['rs-storage']['device']['detach_timeout']` - Amount of time (in seconds) to wait for a volume to detach during
  the decommission of the server. Default is `300` (5 minutes).
- `node['rs-storage']['device']['destroy_on_decommission']` - Whether to destroy the device during the decommission of
  the server. Default is `false`.
- `node['rs-storage']['device']['mkfs_options']` - Additional mkfs options for formatting the device. Default is `'-F'`
  which is required to avoid warnings about formatting the whole device.
- `node['rs-storage']['device']['stripe_size']` - The stripe size to use on LVM. Default is `512`.
- `node['rs-storage']['backup']['lineage']` - The backup lineage. Default is `nil`.
- `node['rs-storage']['backup']['keep']['keep_last']` - Maximum snapshots to keep. Default is `60`.
- `node['rs-storage']['backup']['keep']['dailies']` - Number of daily backups to keep. Default is `14`.
- `node['rs-storage']['backup']['keep']['weeklies']` - Number of weekly backups to keep. Default is `6`.
- `node['rs-storage']['backup']['keep']['monthlies']` - Number of monthly backups to keep. Default is `12`.
- `node['rs-storage']['backup']['keep']['yearlies']` - Number of yearly backups to keep. Default is `2`.
- `node['rs-storage']['restore']['lineage']` - The name of the lineage to restore the backups from. Default is `nil`.
- `node['rs-storage']['restore']['timestamp']` - The timestamp to restore backup taken on or before the timestamp in the
  same lineage. Default is `nil`.
- `node['rs-storage']['schedule']['enable']` - Enable/disable automated backups. Default is `false`.
- `node['rs-storage']['schedule']['hour']` - The backup schedule hour. Default is `nil`.
- `node['rs-storage']['schedule']['minute']` - The backup schedule minute. Default is `nil`.


# Recipes

## `rs-storage::default`

Simply includes the `rightscale_volume::default` and `rightscale_backup::default` recipes to meet the requirements of
using the resources in these cookbooks.

## `rs-storage::volume`

Creates a new volume from scratch or from an existing backup based on the value provided in the
`node['rs-storage']['restore']['lineage']` attribute. If this attribute is set, the volume will be restored from a
backup matching this lineage otherwise a new volume will be created from scratch. This recipe will also format the
volume using the filesystem specified in `node['rs-storage']['device']['filesystem']` and mount the volume on the
location specified in `node['rs-storage']['device']['mount_point']`.

## `rs-storage::stripe`

Creates a new logical volume composed of volumes from scratch or from an existing backup based on the value provided in
the `node['rs-storage']['restore']['lineage']` attribute. If this attribute is set, the volumes will be restored from a
backup matching this lineage otherwise a new logical volume composed of volumes will be created from scratch. This
recipe will create a striped logical volume using LVM on the volumes and format the logical volume using the filesystem
specified in `node['rs-storage']['device']['filesystem']`. This will also mount the volume on the location specified in
`node['rs-storage']['device']['mount_point']`.

## `rs-storage::backup`

Takes a backup of all volumes attached to the server (except boot disks if there are any) with the lineage specified
in the `node['rs-storage']['backup']['lineage']` attribute. During the backup process, the filesystem will be frozen.i
The filesystem will be unfrozen even if the backup process fails with the help of a chef exception handler. This recipe
will also clean up the volume snapshots based on the criteria specified in the `rs-storage/backup/keep/*` attributes.

## `rs-storage::schedule`

Adds or removes the crontab entry for taking backups periodically at the minute and hour provided via
`node['rs-storage']['schedule']['minute']` and `node['rs-storage']['schedule']['hour']` attributes. The recipe uses the
`node['rs-storage']['schedule']['enable']` attribute to determine wheter to add or remove the crontab entry.

## `rs-storage::decommission`

If the `node['rs-storage']['device']['destroy_on_decommission']` attribute is set to true, this recipe detaches and
deletes the volumes attached to the server. This operation will be skipped if the server is entering the stop state or
rebooting.

# Author

Author:: RightScale, Inc. (<cookbooks@rightscale.com>)


CHANGELOG.md
rs-storage Cookbook CHANGELOG
=============================

This file is used to list changes made in each version of the rs-storage cookbook.

v2.0.1
------
- loosened lvm lock, so that it can keep up with lvm changes.
- updates filesystem cookbook to >= 0.11.1 to use the supermarket version and \
  fix the frozen? issue.

v2.0.0
------

- Add support for Chef 12
- Remove support for Chef 11

v1.2.1
------
- pin Gems and cookbooks using Gemfile.lock and Berksfile.lock


v1.2.0
------

- updated lvm cookbook to 1.6.1

v1.1.0
------

- Support for RL10

v1.0.7
------

- Issue https://github.com/gregsymons/di-ruby-lvm-attrib/issues/22 has been fixed so workaround can be removed.

v1.0.6
------

- On RHEL, depending on cloud, check and wait for RHEL repos to be installed.

v1.0.5
------

- Handle use of XFS when using rs-storage::volume.

v1.0.4
------

- Due to issue https://github.com/gregsymons/di-ruby-lvm-attrib/issues/22, updated gem added to cookbook
  to allow use with RHEL/CentOS 7.1.
- Updated lvm cookbook dependency version to 1.3.6.

v1.0.3
------

- Fix volume type input description since it is not just for vSphere.

v1.0.2
------

- Require the latest [rightscale_volume (v1.2.4)] and [rightscale_backup (v1.1.5)] cookbooks.

[rightscale_volume (v1.2.4)]: https://github.com/rightscale-cookbooks/rightscale_volume/releases/tag/v1.2.4
[rightscale_backup (v1.1.5)]: https://github.com/rightscale-cookbooks/rightscale_backup/releases/tag/v1.1.5

v1.0.1
------

- `rs-storage/device/controller_type` node attribute created and can be passed as an option
  to `rightscale_volume` resource.

v1.0.0
------

- Initial release


metadata.rb
name             'rs-storage'
maintainer       'RightScale, Inc.'
maintainer_email 'cookbooks@rightscale.com'
license          'Apache 2.0'
description      'Provides recipes for managing volumes on a Server in a RightScale supported cloud'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '2.0.1'
issues_url       'https://github.com/rightscale-cookbooks/rs-storage/issues' if respond_to?(:issues_url)
source_url       'https://github.com/rightscale-cookbooks/rs-storage' if respond_to?(:source_url)
chef_version '>= 12.0' if respond_to?(:chef_version)

depends 'chef_handler'
depends 'filesystem', '>= 0.11.1'
depends 'lvm', '>= 3.0'
depends 'marker'
depends 'rightscale_backup', '~> 2.0.0'
depends 'rightscale_volume', '~> 2.0.0'
depends 'apt'
depends 'build-essential'

recipe 'rs-storage::default', 'Sets up required dependencies for using this cookbook'
recipe 'rs-storage::volume', 'Creates a volume and attaches it to the server'
recipe 'rs-storage::stripe', 'Creates volumes, attaches them to the server, and sets up a striped LVM'
recipe 'rs-storage::backup', description: 'Creates a backup', thread: 'storage_backup'
recipe 'rs-storage::decommission', 'Destroys LVM conditionally, detaches and destroys volumes. This recipe should' \
                                   ' be used as a decommission recipe in a RightScale ServerTemplate.'
recipe 'rs-storage::schedule', 'Enable/disable periodic backups based on rs-storage/schedule/enable'

attribute 'rs-storage/device/count',
  display_name: 'Device Count',
  description: 'The number of devices to create and use in the Logical Volume. If this value is set to more than' \
                  ' 1, it will create the specified number of devices and create an LVM on the devices.',
  default: '2',
  recipes: ['rs-storage::stripe', 'rs-storage::decommission'],
  required: 'recommended'

attribute 'rs-storage/device/mount_point',
  display_name: 'Device Mount Point',
  description: 'The mount point to mount the device on. Example: /mnt/storage',
  default: '/mnt/storage',
  recipes: ['rs-storage::volume', 'rs-storage::stripe', 'rs-storage::decommission'],
  required: 'recommended'

attribute 'rs-storage/device/nickname',
  display_name: 'Device Nickname',
  description: 'Nickname for the device. rs-storage::volume uses this for the filesystem label, which is' \
                  ' restricted to 12 characters.  If longer than 12 characters, the filesystem label will be set to the' \
                  ' first 12 characters. Example: data_storage',
  default: 'data_storage',
  recipes: ['rs-storage::volume', 'rs-storage::stripe', 'rs-storage::decommission'],
  required: 'recommended'

attribute 'rs-storage/device/volume_size',
  display_name: 'Device Volume Size',
  description: 'Size of the volume or logical volume to create (in GB). Example: 10',
  default: '10',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'recommended'

attribute 'rs-storage/device/iops',
  display_name: 'Device IOPS',
  description: 'IO Operations Per Second to use for the device. Currently this value is only used on AWS clouds.' \
                  ' Example: 100',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'optional'

attribute 'rs-storage/device/volume_type',
  display_name: 'Volume Type',
  description: 'Volume Type to use for creating volumes. Example: gp2',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'optional'

attribute 'rs-storage/device/filesystem',
  display_name: 'Device Filesystem',
  description: 'The filesystem to be used on the device. Defaults are based on OS and determined in' \
                  ' attributes/defaults.rb. Example: ext4',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'optional'

attribute 'rs-storage/device/detach_timeout',
  display_name: 'Detach Timeout',
  description: 'Amount of time (in seconds) to wait for a single volume to detach at decommission. Example: 300',
  default: '300',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'optional'

attribute 'rs-storage/device/destroy_on_decommission',
  display_name: 'Destroy on Decommission',
  description: 'If set to true, the devices will be destroyed on decommission.',
  default: 'false',
  recipes: ['rs-storage::decommission'],
  required: 'recommended'

attribute 'rs-storage/backup/lineage',
  display_name: 'Backup Lineage',
  description: 'The backup lineage. Example: production',
  recipes: ['rs-storage::backup'],
  required: 'required'

attribute 'rs-storage/restore/lineage',
  display_name: 'Restore Lineage',
  description: 'The lineage name to restore backups. Example: staging',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'recommended'

attribute 'rs-storage/restore/timestamp',
  display_name: 'Restore Timestamp',
  description: 'The timestamp (in seconds since UNIX epoch) to select a backup to restore from.' \
                  ' The backup selected will have been created on or before this timestamp. Example: 1391473172',
  recipes: ['rs-storage::volume', 'rs-storage::stripe'],
  required: 'recommended'

attribute 'rs-storage/backup/keep/dailies',
  display_name: 'Backup Keep Dailies',
  description: 'Number of daily backups to keep. Example: 14',
  default: '14',
  recipes: ['rs-storage::backup'],
  required: 'optional'

attribute 'rs-storage/backup/keep/weeklies',
  display_name: 'Backup Keep Weeklies',
  description: 'Number of weekly backups to keep. Example: 6',
  default: '6',
  recipes: ['rs-storage::backup'],
  required: 'optional'

attribute 'rs-storage/backup/keep/monthlies',
  display_name: 'Backup Keep Monthlies',
  description: 'Number of monthly backups to keep. Example: 12',
  default: '12',
  recipes: ['rs-storage::backup'],
  required: 'optional'

attribute 'rs-storage/backup/keep/yearlies',
  display_name: 'Backup Keep Yearlies',
  description: 'Number of yearly backups to keep. Example: 2',
  default: '2',
  recipes: ['rs-storage::backup'],
  required: 'optional'

attribute 'rs-storage/backup/keep/keep_last',
  display_name: 'Backup Keep Last Snapshots',
  description: 'Number of snapshots to keep. Example: 60',
  default: '60',
  recipes: ['rs-storage::backup'],
  required: 'optional'

attribute 'rs-storage/schedule/enable',
  display_name: 'Backup Schedule Enable',
  description: 'Enable or disable periodic backup schedule',
  default: 'false',
  choice: %w(true false),
  recipes: ['rs-storage::schedule'],
  required: 'recommended'

attribute 'rs-storage/schedule/hour',
  display_name: 'Backup Schedule Hour',
  description: "The hour to schedule the backup on. This value should abide by crontab syntax. Use '*' for taking' +
    ' backups every hour. Example: 23",
  recipes: ['rs-storage::schedule'],
  required: 'required'

attribute 'rs-storage/schedule/minute',
  display_name: 'Backup Schedule Minute',
  description: 'The minute to schedule the backup on. This value should abide by crontab syntax. Example: 30',
  recipes: ['rs-storage::schedule'],
  required: 'required'
